crates/cranelift-shared/src/compiled_function.rs

```diff
@@ -0,0 +1,86 @@
+use crate::{mach_reloc_to_reloc, mach_trap_to_trap, Relocation};
+use cranelift_codegen::{
+    ir, ir::UserExternalNameRef, isa::unwind::UnwindInfo, Final, MachBufferFinalized,
+    ValueLabelsRanges,
+};
+use wasmtime_environ::{FilePos, InstructionAddressMap, TrapInformation};
+
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+/// Metadata to translate from binary offsets back to the original
+/// location found in the wasm input.
+pub struct FunctionAddressMap {
+    /// An array of data for the instructions in this function, indicating where
+    /// each instruction maps back to in the original function.
+    ///
+    /// This array is sorted least-to-greatest by the `code_offset` field.
+    /// Additionally the span of each `InstructionAddressMap` is implicitly the
+    /// gap between it and the next item in the array.
+    pub instructions: Box<[InstructionAddressMap]>,
+
+    /// Function's initial offset in the source file, specified in bytes from
+    /// the front of the file.
+    pub start_srcloc: FilePos,
+
+    /// Function's end offset in the source file, specified in bytes from
+    /// the front of the file.
+    pub end_srcloc: FilePos,
+
+    /// Generated function body offset if applicable, otherwise 0.
+    pub body_offset: usize,
+
+    /// Generated function body length.
+    pub body_len: u32,
+}
+
+/// Compiled function: machine code body, jump table offsets, and unwind information.
+#[derive(Default)]
+pub struct CompiledFunction {
+    /// The machine code for this function.
+    pub body: Vec<u8>,
+
+    /// The unwind information.
+    pub unwind_info: Option<UnwindInfo>,
+
+    /// Information used to translate from binary offsets back to the original
+    /// location found in the wasm input.
+    pub address_map: FunctionAddressMap,
+
+    /// Metadata about traps in this module, mapping code offsets to the trap
+    /// that they may cause.
+    pub traps: Vec<TrapInformation>,
+
+    pub relocations: Vec<Relocation>,
+    pub value_labels_ranges: ValueLabelsRanges,
+    pub sized_stack_slots: ir::StackSlots,
+    pub alignment: u32,
+}
```

Comment:
One thing I attempted a long time ago but didn't get a chance to finish, which I think might simplify the winch/cranelift duo in the long run, is to replace this entire structure with `MachBufferFinalized`. Most of these fields are already present within there and the copies here are simple translations from Cranelift datatypes to `wasmtime-cranelift-shared` datatypes. In such a situation I think it'd be better to leave everything at-rest as Cranelift datatypes and only much later, for example when inserting into an object, is anything translated from Cranelift.

Would you be up for exploring such a change like this?

Reply:
Right, I see what you mean. Instead of eagerly converting from `MachBufferFinalized<T>` into the current `CompiledFunction` in `compile_function`, the idea would be to defer any conversions when needed, e.g. in `append_code` for example. 


Now that I'm a bit more acquainted with this part of the codebase I think I can try this out.